### ISSUE TITLE
Unexpected / non-monotonic trends when reproducing the Generalization (Arrival Rate) and Solvability experiments — possible config or metric-definition issue

### ISSUE THREAD

**Issue by Mehrabpsh** [GeminiLight/virne#45]
<h2>Summary</h2>
<p>I'm trying to reproduce the paper's <strong>Solvability</strong> (VN-size sweep) and <strong>Generalization</strong> (arrival-rate sweep) experiments using the provided solvers (<code>ppo_dual_gat+</code>, <code>a3c_dual_gcn+</code>, <code>ppo_mlp+</code>, <code>order_rank</code>). The Solvability results are close to the paper <em>except</em> for a rank swap between two RL solvers, but the Generalization (arrival-rate) results show trends that don't make physical sense to me, in particular a <strong>non-learned heuristic getting <em>more</em> successful as load increases</strong>. I've checked my own analysis/plotting code carefully and don't believe it's the source of the problem, but I want to lay out exactly what I ran and what I'm seeing in case it's a config misunderstanding on my end, a train/test setting mismatch, or an actual bug.</p>
<p>I'm not certain which of the three it is — that's exactly what I'm hoping to get help narrowing down:</p>
<ol>
<li>A bug in my own analysis/plotting scripts.</li>
<li>A misconfiguration on my side / settings not aligned with what the paper used.</li>
<li>A bug or unexpected behavior in Virne itself.</li>
</ol>
<hr>
<h2>Environment</h2>
<ul>
<li>Virne commit: <code>145553e8e39d5ca4de974e5788575bdd8e5fb11d</code></li>
<li>OS: Mint (Linux)</li>
<li>Key packages: <code>torch==2.6.0+cu124</code>, <code>torch-geometric==2.7.0</code>, <code>hydra-core==1.3.2</code>, <code>omegaconf==2.3.0</code>, <code>gym==0.22.0</code>, <code>numpy==2.2.6</code>, <code>pandas==2.3.3</code></li>
<li>Command entry point: <code>main.py</code> (Hydra-based config overrides), <code>system.if_offline_system</code> toggled per experiment</li>
<li>Physical network setting: <code>wx100_p_net_setting</code></li>
</ul>
<hr>
<h2>What I ran</h2>
<p>Before each experiment (and before switching between its training and testing phases), I edited <code>settings/main.yaml</code>'s Hydra <code>defaults</code> list via <code>sed</code>, swapping the generic entries for the simulation-specific config files, e.g.:</p>
<pre><code class="language-yaml">defaults:
  - v_sim_setting/default        # -&gt; v_sim_setting/&lt;simulation_specific&gt;
  - p_net_setting/default        # -&gt; p_net_setting/&lt;simulation_specific&gt;
</code></pre>
<p>Concretely: for Solvability training I pointed <code>v_sim_setting</code> at <code>v_sim_setting_solvability_train</code> and <code>p_net_setting</code> at <code>wx100_p_net_setting</code>; for Solvability testing at <code>v_sim_setting_solvability_test</code>; for Generalization training at <code>v_sim_setting_generalization_arrivalrate_train</code>; for Generalization testing at <code>v_sim_setting_generalization_arrivalrate_test</code> (always <code>wx100_p_net_setting</code> for the physical network). I restored <code>main.yaml</code> to its original defaults after each full experiment finished. The CLI overrides shown below (<code>v_sim_setting.v_net_size.low/high</code>, <code>v_sim_setting.arrival_rate.lam</code>, etc.) are then applied on top of whichever <code>v_sim_setting</code> file was active via this defaults-list swap.</p>
<h3>1. Solvability experiment (VN size sweep, <code>VN_SIZE ∈ [2, 10]</code>)</h3>
<ul>
<li>Train VN config: <code>v_sim_setting_solvability_train</code></li>
<li>Test VN config: <code>v_sim_setting_solvability_test</code></li>
<li><code>TRAIN_SEED=0</code>, <code>TRAIN_EPOCHS=50</code>, <code>training.save_interval=1</code> (for RL solvers only: <code>ppo_dual_gat+</code>, <code>a3c_dual_gcn+</code>, <code>ppo_mlp+</code>)</li>
</ul>
<p>Training command:</p>
<pre><code class="language-bash">python main.py \
    solver.solver_name="${solver_key}" \
    training.num_train_epochs="${NUM_EPOCHS}" \
    system.if_offline_system=true \
    experiment.seed="${TRAIN_SEED}" \
    experiment.num_simulations=1 \
    experiment.run_id="${solver_key}__solvability_train_seed_${TRAIN_SEED}_num_epochs_${TRAIN_EPOCHS}"
</code></pre>
<p>Evaluation command (RL solvers, using the trained checkpoint; <code>order_rank</code> run the same way minus <code>solver.pretrained_model_path</code>), swept over <code>VN_SIZE ∈ {2,...,10}</code> and <code>TEST_SEED ∈ {0, 1111, 2222, 3333, 4444, 5555, 6666, 7777, 8888, 9999}</code>:</p>
<pre><code class="language-bash">python main.py \
    solver.solver_name="${solver_key}" \
    training.num_train_epochs=0 \
    system.if_offline_system=true \
    solver.pretrained_model_path="${model_path}" \
    v_sim_setting.v_net_size.low="${VN_SIZE}" \
    v_sim_setting.v_net_size.high="${VN_SIZE}" \
    experiment.seed="${TEST_SEED}" \
    experiment.num_simulations=1 \
    experiment.run_id="${solver_key}__solvability_test_seed_${TEST_SEED}_vn_size_${VN_SIZE}"
</code></pre>
<h3>2. Generalization experiment (arrival-rate sweep, <code>λ ∈ [0.04, 0.28]</code>, step 0.04)</h3>
<ul>
<li>Train VN config: <code>v_sim_setting_generalization_arrivalrate_train</code>, <strong>fixed</strong> training λ = 0.16</li>
<li>Test VN config: <code>v_sim_setting_generalization_arrivalrate_test</code></li>
<li>Same <code>TRAIN_SEED</code>/<code>TRAIN_EPOCHS</code> as above</li>
</ul>
<p>Training command:</p>
<pre><code class="language-bash">python main.py \
    solver.solver_name="${solver_key}" \
    training.num_train_epochs="${NUM_EPOCHS}" \
    experiment.seed="${TRAIN_SEED}" \
    experiment.num_simulations=1 \
    experiment.run_id="${solver_key}__generalization_AR_train_lambda_${TRAIN_ARRIVAL_RATE}_seed_${TRAIN_SEED}_num_epochs_${TRAIN_EPOCHS}"
</code></pre>
<p>Evaluation command, swept over <code>λ ∈ {0.04, 0.08, ..., 0.28}</code> and the same 10 <code>TEST_SEED</code> values:</p>
<pre><code class="language-bash">python main.py \
    solver.solver_name="${solver_key}" \
    training.num_train_epochs=0 \
    solver.pretrained_model_path="${model_path}" \
    v_sim_setting.arrival_rate.lam="${lambda}" \
    experiment.seed="${TEST_SEED}" \
    experiment.num_simulations=1 \
    experiment.run_id="${solver_key}__generalization_AR_test_lambda_${lambda}_seed_${TEST_SEED}"
</code></pre>
<p>(<code>order_rank</code> run identically minus <code>solver.pretrained_model_path</code>.)</p>
<p><strong>Note the asymmetry:</strong> the Solvability commands explicitly set <code>system.if_offline_system=true</code> on both training and evaluation; the Generalization commands never set <code>system.if_offline_system</code> at all, so they run under whatever the default resolves to (I believe <code>false</code>, i.e. the online/event-driven system, but haven't confirmed this in the source). Solvability's rank ordering looks basically correct; Generalization's doesn't. I'm not sure whether this difference matters, but flagging it in case it's relevant.</p>
<h3>3. Analysis / plotting</h3>
<p>For each experiment I locate every result directory by regex-matching <code>run_id</code> against the pattern above, and for each <code>(solver, λ_or_size, test_seed)</code> triple I read the <code>acceptance_rate</code>, <code>long_term_time_r2c_ratio</code>, and <code>long_term_avg_revenue</code> columns out of that run's <code>summary.csv</code>. I then group by <code>(solver, λ_or_size)</code> across the 10 <code>test_seed</code> values, average each metric column with <code>np.mean</code> (and compute <code>np.std</code> for error bars), and build a new dataframe indexed by <code>model</code> and <code>λ</code>/<code>size</code> only (no <code>test_seed</code> column, since it's been averaged out). That dataframe is what gets plotted below.</p>
<hr>
<h2>Observed results</h2>
<h3>Solvability heatmap (avg. over 10 seeds)</h3>
<img width="3303" height="2067" alt="Image" src="https://github.com/user-attachments/assets/1216f410-e5ca-4140-ab51-935f7e0cceca" />
<p>Mostly consistent with the paper's ordering, <strong>except</strong>:</p>

VN size | PPO-DUAL-GAT+ | A3C-DUAL-GCN+
-- | -- | --
2 | 1.000 | 0.983
3 | 0.970 | 0.962
4–10 | ~equal, A3C slightly ahead at larger sizes (e.g. size 10: 0.698 vs 0.706) |  

<p>In the paper, PPO-DUAL-GAT+ is reported to outperform A3C-DUAL-GCN+ consistently. Here A3C is essentially tied or slightly <em>ahead</em> at larger VN sizes. This could plausibly be explained by training-budget/hyperparameter differences from the paper, so I'm flagging it but it's not the main concern.</p>
<h3>Generalization (arrival rate)</h3>
<img width="4464" height="3069" alt="Image" src="https://github.com/user-attachments/assets/664a1dbe-3d14-4e45-b044-1040bd75c18f" />
<p>Three things look wrong to me:</p>
<ol>
<li><strong>Acceptance rate vs λ</strong>: For all three RL solvers, acceptance rate decreases as λ increases (expected — higher arrival rate → more contention), but the <em>rate of decrease shrinks</em> as λ grows, and for A3C-DUAL-GCN+ it even dips and partially recovers (λ=0.16→0.20→0.24 goes 0.78→0.75→0.78). Not a clean monotonic relationship.</li>
<li><strong>ORDER-RANK acceptance rate <em>increases</em> with λ</strong> (from ~0.48 at λ=0.04 up to ~0.62 at λ=0.24, dipping slightly at λ=0.28). This is the most concerning result: ORDER-RANK is a fixed, non-learned heuristic — it has no training distribution to generalize (in)correctly from, so its acceptance rate should be a <strong>clean, monotonic function of the actual load the simulator produces</strong>. Seeing it improve as configured λ increases strongly suggests to me that the <em>realized</em> arrival rate is not increasing the way <code>λ</code> is intended to.</li>
<li><strong><code>long_term_time_r2c_ratio</code> increases with λ for every solver, including ORDER-RANK.</strong> I'd expect solution quality (R2C ratio) to hold steady or degrade under higher contention, not improve for all four methods simultaneously.</li>
<li>Error bars (std over 10 seeds) for ORDER-RANK at low λ are very large (e.g., acceptance rate 0.48 ± ~0.25 at λ=0.04), shrinking substantially at higher λ — suggesting possible small-sample variance in the number of arrivals generated at low λ, on top of the trend issue above.</li>
</ol>
<p>All three of these point in the same direction: <strong>it looks like actual simulated load may be behaving inversely to the configured <code>λ</code></strong>, rather than a plotting or aggregation bug on my end (since the aggregation logic is identical between the Solvability and Generalization scripts, and Solvability results look sane).</p>
<hr>
<h2>Did I do something wrong?</h2>
<p>I've gone through my own scripts multiple times and can't find a bug in the commands, the seed/config sweeps, or the aggregation/plotting logic — but I also can't rule out that I'm missing something about how the Generalization experiment is supposed to be configured (e.g. the <code>system.if_offline_system</code> asymmetry noted above, or some other field I should be overriding alongside <code>arrival_rate.lam</code>). If it would help, I'm happy to provide more detail on any part of this — the full driver scripts, the exact YAML contents of the VN configs I used, raw <code>summary.csv</code> files, per-seed (non-averaged) data, or anything else that would make it easier to tell whether this is a setup issue on my end or something worth looking into in Virne itself.</p>
<hr>
<h2>Reproduction</h2>
<p>The exact <code>main.py</code> commands for training and evaluation in both experiments are given above. Full driver scripts (which just loop those commands over solvers/seeds/λ or VN-size, with resume-on-rerun logic) and the analysis/plotting code are attached:</p>
<ul>
<li><code>run_solvability.sh</code>, <code>run_generalization_ar.sh</code></li>
<li><code>utils.py</code>, <code>plot_solvability.py</code>, <code>plot_generalization_ar.py</code></li>
<li><code>solvability_heatmap_avg.png</code>, <code>generalization_ar_metrics_avg.png</code> (embedded above)</li>
</ul>
<p>Happy to run any diagnostic you'd like (e.g. instrumenting the VN request generator to log realized inter-arrival times per λ under both <code>if_offline_system</code> settings) and report back.</p></body></html>

**Comment by GeminiLight** [comment]
Hi Mehrabpsh,

Thanks for the detailed report. 

1. In the offline solvability study, we set `if_offline_system=1`. The paper compares PPO-DualGAT with PPO-DualGCN, while your experiment uses A3C-DualGCN, so that the ranking result may change.

2. In the generalization study on the arrival rate, we set `if_offline_system=0` to simulate the online environment. Could you check whether this parameter is 0 in this experiment?

**Comment by Mehrabpsh** [comment]
Hi,

1. For the solvability experiment, I'll rerun with **PPO-DualGCN+** instead of A3C-DualGCN+ to match the paper's comparison and report back.

2. For the generalization (arrival rate) experiment, `if_offline_system` is already `false` — both in `main.yaml`:
```yaml
system:
  renew_v_net_simulator: false
  if_offline_system: false
  if_changeable_v_nets: false
  if_time_window: false
  time_window_size: 100
```
and in the runtime-printed config:
```
'system': {'if_changeable_v_nets': False,
            'if_offline_system': False,
            'if_time_window': False,
            'renew_v_net_simulator': False,
            'time_window_size': 100},
```
So the generalization experiment is already running in online VNE mode as intended.

Given that, the `if_offline_system` setting doesn't seem to explain the acceptance-rate/R2C trends I'm seeing (especially ORDER-RANK's acceptance rate increasing with λ, since it has no training distribution to generalize incorrectly from). Do you have any other ideas of what could cause that, or specific things you'd like me to check? 
Do you have any script of your simulation so i check to see if mine aligns with it?
